linc_convert/utils/zarr/generate_pyramid.py

```diff
@@ -0,0 +1,165 @@
+        prev_shape, shape = shape, []
+        for i, length in enumerate(prev_shape):
+            if i == no_pyramid_axis:
+                shape.append(length)
+            else:
+                shape.append(max(1, length // 2))
+
+        # Stop if seen enough levels or if the level's shape is smaller than the chunk size
+        if levels == -1:
+            if all(x <= c for x, c in zip(shape, chunk_size[-ndim:])):
+                break
+        elif level > levels:
+            break
+
+        print("Compute level", level, "with shape", shape)
+
+        allshapes.append(shape)
+        omz.create_dataset(str(level), shape=batch + shape, **opt)
+
+        # Iterate across `max_load` chunks
+        # (note that these are unrelated to underlying zarr chunks)
+        grid_shape = [ceildiv(n, max_load) for n in prev_shape]
+        for chunk_index in itertools.product(*[range(x) for x in grid_shape]):
+            print(f"chunk {chunk_index} / {tuple(grid_shape)})", end="\r")
+
+            # Read one chunk of data at the previous resolution
+            slicer = [Ellipsis] + [
+                slice(i * max_load, min((i + 1) * max_load, n))
+                for i, n in zip(chunk_index, prev_shape)
+            ]
+            fullshape = omz[str(level - 1)].shape
+            dat = omz[str(level - 1)][tuple(slicer)]
+
+            # Discard the last voxel along odd dimensions
+            crop = [
+                0 if y == 1 else x % 2 for x, y in zip(dat.shape[-ndim:], fullshape[-ndim:])
+            ]
+            # Only crop the axes that are downsampled
+            if no_pyramid_axis is not None:
+                crop[no_pyramid_axis] = 0
+            slcr = [slice(-1) if x else slice(None) for x in crop]
+            dat = dat[tuple([Ellipsis, *slcr])]
+
+            if any(n == 0 for n in dat.shape):
+                # last strip had a single voxel, nothing to do
+                continue
+
+            patch_shape = dat.shape[-ndim:]
+
+            # Reshape into patches of shape 2x2x2
+            windowed_shape = [
+                x for n in patch_shape for x in (max(n // 2, 1), min(n, 2))
+            ]
+            if no_pyramid_axis is not None:
+                windowed_shape[2 * no_pyramid_axis] = patch_shape[no_pyramid_axis]
+                windowed_shape[2 * no_pyramid_axis + 1] = 1
+
+            dat = dat.reshape(batch + windowed_shape)
+            # -> last `ndim` dimensions have shape 2x2x2
+            dat = dat.transpose(
+                list(range(len(batch)))
+                + list(range(len(batch), len(batch) + 2 * ndim, 2))
+                + list(range(len(batch) + 1, len(batch) + 2 * ndim, 2))
+            )
+            # -> flatten patches
+            smaller_shape = [max(n // 2, 1) for n in patch_shape]
+            if no_pyramid_axis is not None:
+                smaller_shape[no_pyramid_axis] = patch_shape[no_pyramid_axis]
+
+            dat = dat.reshape(batch + smaller_shape + [-1])
```

Comment:
I might be misunderstanding, but I am not sure why this flattening is needed since the data is already reshaped with 2x2x2 patches in the previous section.